### PR TITLE
ignore impossible branches

### PR DESCRIPTION
**Pull request by skogsbaer**
This PR improve on PR #98 by fixing #36 und #56 in a more general way. Compared with #98, the following function now also type checks successfully:

~~~erlang
-spec g(a) -> c; (b) -> d.
g(a) -> c;
g(b) -> d.

-spec f4(a) -> 1; (b) -> 2.
f4(X) ->
    case g(X) of
        c -> 1;
        d -> 2
    end.
~~~

With PR #98, the first (second) branch in f4 are not ignored when type checking against `(b) -> 2` (`(a) -> 1`). 

Further, this PR makes constraint simplification much easier because we no longer need to apply the result of tally during constraint simplification. Instead, we only use tally for checking satisifiability of a constraint to ignore certain branches.

### CODE REVIEW

**Comment by skogsbaer** [comment]
There are still a couple of dialyzer issues.

Further, I had to disable one test case because of bug #101.

**Comment by skogsbaer** [comment]
@albsch ready to review

**Comment by albsch** [comment]
>  Instead, we only use tally for checking satisfiability of a constraint to ignore certain branches.

You mean checking a constraint set for satisfiability, right? This means tally could stop after finding the first solution as an optimization.

> Compared with https://github.com/etylizer/etylizer/pull/98

Do we still need #98?

File: Makefile [hunk 1]
```diff
@@ -2,7 +2,7 @@
 	check_syntax_antidote check_ast_trans_antidote check_types_antidote \
 	check_syntax_riak check_ast_trans_riak check_types_riak
 
-REBAR = ./rebar3
+REBAR = rebar3
```
Comment (albsch) [inline Makefile:5]: We should change this outside of this PR

File: src/ast.erl
Comment (albsch) [inline src/ast.erl:1]: Could you roughly explain in the PR what the loc changes are used for?

File: src/erlang_types/ty_variable.erl [hunk 2]
```diff
@@ -9,6 +9,8 @@
 
 -export([new/1, smallest/3, normalize/6, transform/2, get_new_id/0]).
 
+-export_type([var/0]).
```
Comment (albsch) [inline src/erlang_types/ty_variable.erl:12]: Where is this type used?

File: test_files/tycheck/type_error_msgs/invalid_arg.erl [hunk 3]
```diff
@@ -1,5 +1,5 @@
 % ERROR
-% test_files/tycheck/type_error_msgs/invalid_arg.erl:13:1: Type error: function make_even/1 failed to type check against type fun((integer()) -> integer())
+% test_files/tycheck/type_error_msgs/invalid_arg.erl:13:17: Type error: expression failed to type check
```
Comment (albsch) [inline test_files/tycheck/type_error_msgs/invalid_arg.erl:2]: Do we lose information for the new error messages?

**Review by albsch — CHANGES_REQUESTED** [review]

**Comment by albsch** [comment]
Some old tests are now faster, but this one fails now:

```erlang
-spec fun_local_03() -> integer().
fun_local_03() ->
    F = fun
            Add(0) -> 0;
            Add(X) -> X + Add(X + 1)
        end,
    F(3).
```

The reason is that 

```
test_files/tycheck_simple.erl:432:13: Type error: this branch never matches
%  432|             Add(0) -> 0;
%     |             ^
  while checking function fun_local_03/0 against type fun(() -> integer())
```

Is this intended @skogsbaer ?

File: Makefile (hunk 1 above)
Comment (albsch) [reply Makefile:5]: #102

File: src/erlang_types/ty_variable.erl (hunk 2 above)
Comment (skogsbaer) [reply src/erlang_types/ty_variable.erl:12]: In src/ast_lib.erl. (I did not introduce the use there, but I had to fix it, I guess, for Erlang 26)

File: test_files/tycheck/type_error_msgs/invalid_arg.erl (hunk 3 above)
Comment (skogsbaer) [reply test_files/tycheck/type_error_msgs/invalid_arg.erl:2]: No. In fact, several error message became more precise.

**Comment by skogsbaer** [comment]
> > Instead, we only use tally for checking satisfiability of a constraint to ignore certain branches.
> 
> You mean checking a constraint set for satisfiability, right? This means tally could stop after finding the first solution as an optimization.

Yes! That looks like a very good optimization. I now use `tally:is_satisfiable` at all places where I do not need the resulting substitution but only the check for satisfiability. `tally:is_satisfiable` still returns the first substitution, but this is only for debugging purpose. Currently, `tally:is_satisfiable` is not optimized but still computes all substitutions. But it's now easy to change.

> > Compared with #98
> 
> Do we still need #98?

No, no need for #98.



> One of the previously slow tests fails now:
> 
> ```erlang
> -spec fun_local_03() -> integer().
> fun_local_03() ->
>     F = fun
>             Add(0) -> 0;
>             Add(X) -> X + Add(X + 1)
>         end,
>     F(3).
> ```
> 
> The reason is that
> 
> ```
> test_files/tycheck_simple.erl:432:13: Type error: this branch never matches
> %  432|             Add(0) -> 0;
> %     |             ^
>   while checking function fun_local_03/0 against type fun(() -> integer())

That's indeed still a problem. See my last comment on #36. The fix for it is quite a bit of work. Thus, I propose to merge this PR regardless of the bug. This PR makes things much better than before, even if it still has this flaw. Also we then can continue with merging the PR for cleaning and some other PRs.

**Review by albsch — APPROVED** [review]

**Pull request merged**